lib/ansible/modules/cloud/amazon/aws_glacier_vault.py

```diff
@@ -0,0 +1,227 @@
+                return v
+
+        return None
+
+    def create_or_update(self):
+        vault = self.exists(self.name)
+        changed = False
+        if vault is None:
+            vault = self.connection.create_vault(vaultName=self.name)
+            changed = True
+
+        changed |= self.manage_tags()
+
+        # Vault exists, no change to do
+        self.module.exit_json(vault=vault, changed=changed)
+
+    def list_tags(self):
+        tags = self.connection.list_tags_for_vault(vaultName=self.name)
+        if "Tags" not in tags:
+            self.module.fail_json(msg="Invalid response from Glacier: no Tags found in vault object {0}".format(tags))
+        return tags["Tags"]
+
+    def manage_tags(self):
+        changed = False
+        old_tags = self.list_tags()
+        tags_to_set, tags_to_delete = compare_aws_tags(
+            old_tags, self.module.params.get('tags') or {},
+        )
+        if tags_to_set:
+            self.connection.add_tags_to_vault(
+                vaultName=self.name,
+                Tags=tags_to_set)
+            changed |= True
+        if tags_to_delete:
+            self.connection.remove_tags_from_vault(
+                vaultName=self.name,
+                TagKeys=tags_to_delete)
+            changed |= True
+        return changed
+
+    def destroy(self):
+        name = self.module.params.get('name')
+        vault = self.exists(name)
+        if vault is None:
+            self.module.exit_json(changed=False)
+
+        self.connection.delete_vault(vaultName=name)
+        self.module.exit_json(vault=vault, changed=True)
+
+
+def main():
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(
+        dict(
+            name=dict(required=True, type='str'),
+            state=dict(default='present', type='str', choices=['present', 'absent']),
+            tags=dict(type='dict'),
+        )
+    )
+
+    module = AnsibleModule(argument_spec=argument_spec)
+
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 required for this module')
+
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
+
+    if region:
+        connection = boto3_conn(
+            module,
+            conn_type='client',
+            resource='glacier',
+            region=region,
+            endpoint=ec2_url,
+            **aws_connect_params
+        )
+    else:
+        module.fail_json(msg="region must be specified")
+
+    if connection is None:  # this should never happen
+        module.fail_json(msg='Unknown error, failed to create glacier connection, no information from boto.')
+
+    state = module.params.get("state")
+
+    try:
+        if state == 'present':
+            GlacierVaultManager(connection, module).create_or_update()
+        elif state == 'absent':
+            GlacierVaultManager(connection, module).destroy()
+    except botocore.exceptions.NoCredentialsError as e:
+        module.fail_json(msg="AWS credential error: {0}".format(e))
```

Comment:
use AnsibleAWSModule from module_utils.aws.core, and then use fail_json_aws(e, msg="description of what was attempted")